File: tobas_core/tobas_rqt_tools/src/tobas_rqt_tools/widgets/slider_display.py
import random
from typing import Callable
from PyQt5.QtCore import Qt, pyqtSignal, pyqtSlot, QObject
from PyQt5.QtWidgets import QWidget, QLabel, QLineEdit, QVBoxLayout, QHBoxLayout
from PyQt5.QtGui import QFont

from .slider import Slider, FloatSlider


class IntSliderDisplay(QWidget):
    """整数スライダーとその値の表示機能を持つウィジェット．"""

    PSIZE = 9

    value_changed = pyqtSignal(int)

    def __init__(self) -> None:
        super().__init__()
        self._suffix = ""

        font = QFont("Default", self.PSIZE, QFont.Bold)

        rows = QVBoxLayout()
        self.setLayout(rows)

        cols = QHBoxLayout()
        rows.addLayout(cols)

        self._text = QLabel()
        self._text.setFont(font)
        cols.addWidget(self._text)

        self._value = QLineEdit()
        self._value.setAlignment(Qt.AlignRight)
        self._value.setFont(font)
        self._value.setReadOnly(True)
        self._value.setFocusPolicy(Qt.NoFocus)
        cols.addWidget(self._value)

        self._slider = Slider(Qt.Horizontal)
        rows.addWidget(self._slider)

        self.update()
        self._slider.valueChanged.connect(self._on_value_changed)

    def update(self) -> None:
        self._value.setText(f"{self.get_value()}{self._suffix}")
        self.value_changed.emit(self.get_value())

    def set_text(self, text: str) -> None:
        self._text.setText(text)

    def set_minimum(self, minimum: int) -> None:
        self._slider.setMinimum(minimum)

    def set_maximum(self, maximum: int) -> None:
        self._slider.setMaximum(maximum)

    def get_value(self) -> int:
        return self._slider.value()

    def set_value(self, value: int) -> None:
        self._slider.setValue(value)

    def set_suffix(self, suffix: str) -> None:
        self._suffix = suffix
        self.update()

    def set_callback(self, callback: Callable[[int], None]):
        self.value_changed.connect(callback)

    def set_random_value(self) -> None:
        value = random.randint(self._slider.minimum(), self._slider.maximum())
        self.set_value(value)

    def set_center_value(self) -> None:
        value = (self._slider.minimum() + self._slider.maximum()) // 2
        self.set_value(value)

    @pyqtSlot()
    def _on_value_changed(self) -> None:
        self.update()


class FloatSliderDisplay(QWidget):
    """小数スライダーとその値の表示機能を持つウィジェット．"""

    PSIZE = 9

    value_changed = pyqtSignal(float)

    def __init__(self) -> None:
        super().__init__()
        self._suffix = ""

        font = QFont("Default", self.PSIZE, QFont.Bold)

        rows = QVBoxLayout()
        self.setLayout(rows)

        cols = QHBoxLayout()
        rows.addLayout(cols)

        self._text = QLabel()
        self._text.setFont(font)
        cols.addWidget(self._text)

        self._value = QLineEdit()
        self._value.setAlignment(Qt.AlignRight)
        self._value.setFont(font)
        self._value.setReadOnly(True)
        self._value.setFocusPolicy(Qt.NoFocus)
        cols.addWidget(self._value)

        self._slider = FloatSlider(Qt.Horizontal)
        rows.addWidget(self._slider)

        self.update()
        self._slider.valueChanged.connect(self._on_value_changed)

    def update(self) -> None:
        self._value.setText(f"{self.get_value()}{self._suffix}")
        self.value_changed.emit(self.get_value())

    def set_text(self, text: str) -> None:
        self._text.setText(text)

    def set_minimum(self, minimum: float) -> None:
        self._slider.setMinimum(minimum)

    def set_maximum(self, maximum: float) -> None:
        self._slider.setMaximum(maximum)

    def get_value(self) -> float:
        return self._slider.value()

    def set_value(self, value: float) -> None:
        self._slider.setValue(value)

    def set_suffix(self, suffix: str) -> None:
        self._suffix = suffix

    def set_callback(self, callback: Callable[[float], None]) -> None:
        self.value_changed.connect(callback)

    def set_random_value(self) -> None:
        value = random.uniform(self._slider.minimum(), self._slider.maximum())
        self.set_value(value)

    def set_center_value(self) -> None:
        value = (self._slider.minimum() + self._slider.maximum()) / 2
        self.set_value(value)

    @pyqtSlot()
    def _on_value_changed(self) -> None:
        self.update()

File: tobas_gui/tobas_qt_tools/src/widgets/slider_display.cpp
#include <QVBoxLayout>
#include <QHBoxLayout>

#include "tobas_qt_tools/widgets/slider_display.hpp"
#include "tobas_qt_tools/font.hpp"

#define TEXT_PSIZE 9

namespace qt
{
IntSliderDisplay::IntSliderDisplay(QWidget* parent) : super(parent)
{
  DefaultFont font(TEXT_PSIZE, QFont::Bold);

  const auto rows = new QVBoxLayout();
  setLayout(rows);

  const auto cols = new QHBoxLayout();
  rows->addLayout(cols);

  text_ = new QLabel();
  text_->setFont(font);
  cols->addWidget(text_);

  value_ = new QLineEdit();
  value_->setAlignment(Qt::AlignRight);
  value_->setFont(font);
  value_->setReadOnly(true);
  value_->setFocusPolicy(Qt::NoFocus);
  cols->addWidget(value_);

  slider_ = new Slider(Qt::Horizontal);
  rows->addWidget(slider_);

  connect(slider_, &Slider::valueChanged, this, &self::onSliderValueChanged);
}

int IntSliderDisplay::getValue() const
{
  return slider_->value();
}

void IntSliderDisplay::setValue(int value, bool block_signal)
{
  slider_->blockSignals(true);
  slider_->setValue(value);
  slider_->blockSignals(false);

  value_->setText(QString::number(value) + suffix_);

  if (!block_signal)
    Q_EMIT valueChanged(value);
}

void IntSliderDisplay::setText(const QString& text)
{
  text_->setText(text);
}

void IntSliderDisplay::setMinimum(int minimum)
{
  slider_->setMinimum(minimum);
}

void IntSliderDisplay::setMaximum(int maximum)
{
  slider_->setMaximum(maximum);
}

void IntSliderDisplay::setSuffix(const QString& suffix)
{
  suffix_ = suffix;
  value_->setText(QString::number(getValue()) + suffix_);
}

void IntSliderDisplay::setCenterValue(bool block_signal)
{
  const auto value = (slider_->minimum() + slider_->maximum()) / 2;
  setValue(value, block_signal);
}

void IntSliderDisplay::onSliderValueChanged(int value)
{
  value_->setText(QString::number(value) + suffix_);
  Q_EMIT valueChanged(value);
}

DoubleSliderDisplay::DoubleSliderDisplay(int decimals, QWidget* parent) : super(parent), decimals_(decimals)
{
  DefaultFont font(TEXT_PSIZE, QFont::Bold);

  const auto rows = new QVBoxLayout();
  setLayout(rows);

  const auto cols = new QHBoxLayout();
  rows->addLayout(cols);

  text_ = new QLabel();
  text_->setFont(font);
  cols->addWidget(text_);

  value_ = new QLineEdit();
  value_->setAlignment(Qt::AlignRight);
  value_->setFont(font);
  value_->setReadOnly(true);
  value_->setFocusPolicy(Qt::NoFocus);
  cols->addWidget(value_);

  slider_ = new DoubleSlider(Qt::Horizontal);
  rows->addWidget(slider_);

  connect(slider_, &DoubleSlider::valueChanged, this, &self::onSliderValueChanged);
}

double DoubleSliderDisplay::getValue() const
{
  return slider_->value();
}

void DoubleSliderDisplay::setValue(double value, bool block_signal)
{
  slider_->blockSignals(true);
  slider_->setValue(value);
  slider_->blockSignals(false);

  value_->setText(QString::number(value, 'f', decimals_) + suffix_);

  if (!block_signal)
    Q_EMIT valueChanged(value);
}

void DoubleSliderDisplay::setText(const QString& text)
{
  text_->setText(text);
}

void DoubleSliderDisplay::setMinimum(double minimum)
{
  slider_->setMinimum(minimum);
}

void DoubleSliderDisplay::setMaximum(double maximum)
{
  slider_->setMaximum(maximum);
}

void DoubleSliderDisplay::setSuffix(const QString& suffix)
{
  suffix_ = suffix;
  value_->setText(QString::number(getValue(), 'f', decimals_) + suffix_);
}

void DoubleSliderDisplay::setCenterValue(bool block_signal)
{
  const auto value = (slider_->minimum() + slider_->maximum()) / 2;
  setValue(value, block_signal);
}

void DoubleSliderDisplay::onSliderValueChanged(double value)
{
  value_->setText(QString::number(value, 'f', decimals_) + suffix_);
  Q_EMIT valueChanged(value);
}
}  // namespace qt
